### ISSUE TITLE
Bug: [require-array-sort-compare] false positive when the element type is a conditional type that resolves to string

### ISSUE THREAD

**Issue by im10furry** [typescript-eslint/typescript-eslint#12909]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.
- [X] If I used AI assistance to prepare this issue, I have reviewed and followed the [AI Contribution Policy](https://typescript-eslint.io/contributing/ai-policy/).

### Playground Link

https://typescript-eslint.io/play/#fileType=.ts&code=GYVwdgxgLglg9mABAZzgJygHgCoD4AUAbgIYA2IApsgFyL7aIUAeUFYAJsogAyID8KKGhhgA5olrIhI0QEoA2gF1ZiAN4AoRIhLkqAOlQZ8sgNzqAvuqA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6aRAR1ko9oENo0foVrIA9tDplxAW2KDE6MAG1w2HIiGTIAGnUasG7JEoBzJpMQBlfFSZmAgkJGoMd2IgPYAvt4C66n4%2BQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg10NM8AOXapUAYQAW6aAGsylDul4BfECqA&tokens=false

### Repro Code

```ts
function sort<T>(values: (T extends 0 ? string : string)[]) {
  values.sort();
}
```

### ESLint Config

```js
module.exports = {
  parser: '@typescript-eslint/parser',
  parserOptions: {
    projectService: true,
  },
  rules: {
    '@typescript-eslint/require-array-sort-compare': [
      'error',
      { ignoreStringArrays: true },
    ],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

No error. Every element of `values` is a `string`, so `ignoreStringArrays` should suppress the
report exactly as it does for a plain `string[]`, which is valid in this rule.

### Actual Result

Line 2 reports `Require 'compare' argument  @typescript-eslint/require-array-sort-compare`.

### Additional Info

The element type is a deferred conditional type (`ts.TypeFlags.Conditional` = `1 << 24`), and
`getTypeName` (`packages/type-utils/src/getTypeName.ts`) has no conditional branch — it handles
`StringLike`, type-parameter constraints, all-string unions and string intersections only. So the
element is never recognised as `string`, `ignoreStringArrays` never applies, and the array is then
reported by `isTypeArrayTypeOrUnionOfArrayTypes`.

`checker.getBaseConstraintOfType(elementType)` does resolve it, which is currently unused by either
the string-array check or `getTypeName`.

Inputs with the same behaviour — all report today:

| input | element base constraint | currently |
| --- | --- | --- |
| `(T extends 0 ? string : string)[]` | `string` | reports |
| `(T extends string ? T : never)[]` | `string` | reports |
| `type S = string;` `(T extends 0 ? S : S)[]` | `string` | reports |
| `(T extends 0 ? string : number)[]` | `string \| number` | reports (correct — not a string array) |

For reference, these are already handled and behave as expected: `string[]` (ignored),
`("x" \| "y")[]` (ignored), `String[]` (ignored, one of the cases from #12511),
`T extends string; T[]` (ignored, #12511).

Versions: `@typescript-eslint/eslint-plugin` 8.70.0, `eslint` 10.11.0, `typescript` 5.9.3.

Cross-reference: oxc-project/oxc#26566 reports the same false positive for oxlint's type-aware mode
(`oxlint-tsgolint` 7.0.2002), with identical output — I checked the matrix above against both
implementations and they agree line-for-line on every row, so a fix here would resolve that issue
as well.

**Comment by bradzacher** [comment]
Is this real code that you have written or have seen? Or is it purely hypothetical code?

**Comment by zachkirsch** [comment]
Hey @bradzacher! I am the original reporter of the bug, this is a real problem we are running into and this is the most minimal example I could come up with. Happy to share a more realistic (yet more involved) code sample if that'd be helpful!

**Comment by bradzacher** [comment]
A minimal example is good as a way to reproduce the error without noise. But we're also getting to the point in the project where many bugs reported are 1-in-a-million cases. Also with the advent of agents we also keep getting bug reports for purely hypothetical code that nobody would ever actually write.

So clearer example of the real code would be good to help us understand the impact.

**Comment by zachkirsch** [comment]
@bradzacher here's a slightly more descriptive repro

http://typescript-eslint.io/play/#ts=6.0.2&fileType=.ts&code=JYOwLgpgTgZghgYwgAgKIA9JRHANgSQBMAlAewHdkBvAKGWTgAdHTQwBbCcAfWEIC5kAZzBRQAcwDcdZIzhhgXMLwHDRE6QF8aNQhAS44UFDACuIBAtIgGCAI6ngxgIKEAbsCGkoATwAypAgA1gAUQRA%2BgiJiIOIAlIIAClCk7J4QADxurIQAfNI0cEI%2BFshmFlY2uIFBZORC%2BCAAIhBYaSCeCggA8lB6UBkAwsgQmFyEQsjhPqQwaGPYeER1uSEyfEKCGFg4BCQUANqDALoHxwA0NAnIyanpWTm51DIw3sghCNYiyHzIs8gHAB0wI2x0BXigYBCcTiz3o9Dg5DgwDAtgcTggrg8EP8NRCfDi0no2k0QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6aRAR1ko9oENo0foVrIA9tDplxAW2KDE6MAG1w2HIiGTIAGnUasG7JEoBzJpMQBlfFSZmAgkJGoMd2IgPYAvt4C66n4%2BQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg10NM8AOXapUAYQAW6aAGsylDul4BfECqA&tokens=false

lmk if it's still too abstract and I can paste in our code exactly as it is in the codebase, but it's a bit more involved than this example
